File: 0746-min-cost-climbing-stairs/0746-min-cost-climbing-stairs.cpp
class Solution {
public:
    int minCostClimbingStairs(vector<int>& cost) {
        int n = cost.size();
        vector<int> dp(n+1,0); // we have to reach n or from dp[n] to 0
        dp[n-1] = cost[n-1];
        dp[n-2] = cost[n-2]; // not min as if it has value of n-1, then we can make a wrong jump to n-2 instead of n-3
        // giving wrong answer
        
        
        for(int i=n-3;i>=0;i--)
        {
            dp[i] = min(dp[i+1], dp[i+2]) + cost[i];
        }
        // for(int i:dp)
        //     cout<<i<<" ";
        
        return min(dp[0],dp[1]);
    }
};
/*
this ques is same as climbing steps except you have to minimize the cost here
*/

File: 0746-min-cost-climbing-stairs/0746-min-cost-climbing-stairs.py
class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        dp = [1001 for _ in range(len(cost))]
        l = len(cost)
        dp[l-1] = cost[l-1]
        dp[l-2] = cost[l-2]
        
        # traverse array backwards to 0th element in reverse from n-3
        for i in range(l-3, -1, -1):
            dp[i] = min(dp[i+1],dp[i+2]) + cost[i]
        
        return min(dp[0], dp[1])